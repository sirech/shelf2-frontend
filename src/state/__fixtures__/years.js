const years = () => (
  {
    entities: {
      years: {
        '2010': {
          year: 2010,
          count: 1
        },
        '2012': {
          year: 2012,
          count: 1
        },
        '2013': {
          year: 2013,
          count: 1
        },
        '2014': {
          year: 2014,
          count: 1
        },
        '2016': {
          year: 2016,
          count: 7
        }
      }
    },
    result: [
      2010,
      2012,
      2013,
      2014,
      2016
    ]
  }
)

export default years
