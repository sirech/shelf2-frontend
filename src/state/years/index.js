import * as constants from './constants'
import * as actions from './actions'

export { actions, constants }
export { default as yearsSelector } from './selectors'
export { default as reducer } from './reducers'
