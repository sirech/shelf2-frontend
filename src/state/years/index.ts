import * as constants from './constants'
import * as actions from './slice'

export { actions, constants }
export { default as yearsSelector } from './selectors'
export { default as reducer } from './slice'
