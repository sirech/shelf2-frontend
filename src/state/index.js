import { combineReducers } from 'redux'

import * as R from 'ramda'

import { constants as booksConstants, reducer as booksReducer } from './books'
import {
  constants as searchConstants,
  reducer as searchReducer,
} from './search'
import { constants as yearsConstants, reducer as yearsReducer } from './years'
import { constants as formConstants, reducer as formReducer } from './form'
import { constants as modalConstants, reducer as modalReducer } from './modal'
import { constants as loginConstants, reducer as loginReducer } from './login'

export const actionPicker = list => {
  return actions => {
    const h = R.pick(list)(actions)

    if (Object.keys(h).length !== list.length) {
      throw new Error(
        `Not all requested actions (${list}) exist in the given actions`
      )
    }

    return h
  }
}

const rootReducer = combineReducers({
  [booksConstants.namespace]: booksReducer,
  [searchConstants.namespace]: searchReducer,
  [yearsConstants.namespace]: yearsReducer,
  [formConstants.namespace]: formReducer,
  [modalConstants.namespace]: modalReducer,
  [loginConstants.namespace]: loginReducer,
})

export default rootReducer
