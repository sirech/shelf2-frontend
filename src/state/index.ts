import { ThunkAction, UnknownAction, configureStore } from '@reduxjs/toolkit'

import { constants as booksConstants, reducer as booksReducer } from './books'
import {
  constants as searchConstants,
  reducer as searchReducer,
} from './search'
import { constants as yearsConstants, reducer as yearsReducer } from './years'
import { constants as loginConstants, reducer as loginReducer } from './login'

const store = configureStore({
  reducer: {
    [booksConstants.namespace]: booksReducer,
    [searchConstants.namespace]: searchReducer,
    [yearsConstants.namespace]: yearsReducer,
    [loginConstants.namespace]: loginReducer,
  },
})

export type RootState = ReturnType<typeof store.getState>
export type AppDispatch = typeof store.dispatch
export type AppThunk<ReturnType = void> = ThunkAction<
  ReturnType,
  RootState,
  unknown,
  UnknownAction
>

export default store
