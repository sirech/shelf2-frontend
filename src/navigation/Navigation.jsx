import React, { useEffect } from 'react'
import { useDispatch, useSelector } from 'react-redux'

import { ListGroup } from 'reactstrap'

import NavigationItem from './NavigationItem'

import yearsSelector from './selectors'

import { actions } from 'state/years'

const Navigation = () => {
  const years = useSelector(yearsSelector)
  const dispatch = useDispatch()
  useEffect(() => {
    dispatch(actions.fetchYears())
  }, [dispatch])

  return (
    <ListGroup tag="aside">
      {years.map((entry) => (
        <NavigationItem key={entry.year} {...entry} />
      ))}
    </ListGroup>
  )
}

export default Navigation
