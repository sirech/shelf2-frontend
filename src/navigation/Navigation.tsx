import React, { useEffect } from 'react'

import { ListGroup } from 'reactstrap'

import NavigationItem from './NavigationItem'

import yearsSelector from './selectors'
import { useAppDispatch, useAppSelector } from 'hooks'

import { actions } from 'state/years'

const Navigation = () => {
  const years = useAppSelector(yearsSelector)
  const dispatch = useAppDispatch()
  useEffect(() => {
    dispatch(actions.fetchYears())
  }, [dispatch])

  return (
    <ListGroup tag="aside">
      {years.map((entry) => (
        <NavigationItem key={entry.year} {...entry} />
      ))}
    </ListGroup>
  )
}

export default Navigation
