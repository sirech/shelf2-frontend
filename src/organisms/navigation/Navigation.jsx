// @flow

import React from 'react'
import { connect } from 'react-redux'
import { createStructuredSelector } from 'reselect'

import R from 'ramda'
import { ListGroup } from 'reactstrap'

import NavigationItem from '../../molecules/navigation_item'

import yearsSelector from './selectors'
import { actions } from '../../state/years'

type Props = {
  years: Array<{year: number, count: number}>,
  fetchYears: () => void
}

class Navigation extends React.Component {
  props: Props
  static defaultProps: Props

  componentDidMount () {
    this.props.fetchYears()
  }

  render () {
    const { years } = this.props
    return (
      <ListGroup tag='div'>
        {years.map(entry =>
          <NavigationItem key={entry.year} {...entry} />
        )}
      </ListGroup>
    )
  }
}

Navigation.defaultProps = {
  years: [],
  fetchYears: () => undefined
}

export default connect(
  (state, props) =>
    createStructuredSelector({
      years: yearsSelector
    })(state),
  R.pick(['fetchYears'])(actions)
)(Navigation)
