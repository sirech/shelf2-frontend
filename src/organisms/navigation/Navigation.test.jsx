import React from 'react'

import { fullRender } from 'test'
import state from 'state/__fixtures__'
import Navigation from './Navigation'

// Mock thunk that triggers API request
import { actions } from 'state/years'
jest.mock('state/years/actions', () => ({
  fetchYears: jest.fn(dispatch => jest.fn()),
}))

describe('components', () => {
  describe('Navigation', () => {
    beforeEach(() => {
      jest.clearAllMocks()
    })

    it('renders correctly', () => {
      const { component } = fullRender(<Navigation />, state())
      expect(component.toJSON()).toMatchSnapshot()
    })

    it('fetches the years at the beginning', () => {
      fullRender(<Navigation />, state())
      expect(actions.fetchYears.mock.calls.length).toBe(1)
    })
  })
})
