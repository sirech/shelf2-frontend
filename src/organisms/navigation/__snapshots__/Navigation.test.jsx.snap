// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components Navigation renders correctly 1`] = `
<div
  className="list-group"
>
  <a
    className="list-group-item-action list-group-item"
    href="/books/2010"
    onClick={[Function]}
    style={undefined}
  >
    <span>
      2010
    </span>
    <span
      className="ml-auto badge badge-warning badge-pill"
    >
      1
    </span>
  </a>
  <a
    className="list-group-item-action list-group-item"
    href="/books/2012"
    onClick={[Function]}
    style={undefined}
  >
    <span>
      2012
    </span>
    <span
      className="ml-auto badge badge-warning badge-pill"
    >
      1
    </span>
  </a>
  <a
    className="list-group-item-action list-group-item"
    href="/books/2013"
    onClick={[Function]}
    style={undefined}
  >
    <span>
      2013
    </span>
    <span
      className="ml-auto badge badge-warning badge-pill"
    >
      1
    </span>
  </a>
  <a
    className="list-group-item-action list-group-item"
    href="/books/2014"
    onClick={[Function]}
    style={undefined}
  >
    <span>
      2014
    </span>
    <span
      className="ml-auto badge badge-warning badge-pill"
    >
      1
    </span>
  </a>
  <a
    className="list-group-item-action list-group-item"
    href="/books/2016"
    onClick={[Function]}
    style={undefined}
  >
    <span>
      2016
    </span>
    <span
      className="ml-auto badge badge-warning badge-pill"
    >
      7
    </span>
  </a>
</div>
`;
